code/items/weapons/NewtonLauncher.cs

```diff
@@ -0,0 +1,135 @@
+        public override float ReloadTime => 2.3f;
+        public override float DeployTime => 0.4f;
+        public override int BaseDamage => 3;
+
+        public virtual int Price => 100;
+
+        public static int NEWTON_FORCE_MIN => 300;
+        public static int NEWTON_FORCE_MAX => 700;
+        public static float NEWTON_CHARGE_TIME => 3;
+
+        public bool IsCharging { get; set; } = false;
+        public float ChargingStartTime;
+
+        public override void Spawn()
+        {
+            base.Spawn();
+
+            SetModel("weapons/rust_pistol/rust_pistol.vmdl");
+        }
+
+        public override void OnActive()
+        {
+            base.OnActive();
+            IsCharging = false;
+        }
+
+        public override void ActiveEnd(Entity ent, bool dropped)
+        {
+            base.ActiveEnd(ent, dropped);
+            IsCharging = false;
+        }
+
+        public override void AttackPrimary()
+        {
+            if (IsCharging) return;
+
+            IsCharging = true;
+            ChargingStartTime = Time.Now;
+        }
+
+        public void ReleaseCharge()
+        {
+            (Owner as AnimEntity).SetAnimBool("b_attack", true);
+            ShootEffects();
+            PlaySound("rust_pistol.shoot").SetPosition(Position).SetVolume(0.8f);
+            ShootBullet(0f, 0f, BaseDamage, 0f);
+            IsCharging = false;
+            TimeSincePrimaryAttack = 0;
+        }
+
+        public override void ShootBullet(float spread, float force, float damage, float bulletSize)
+        {
+            Vector3 forward = Owner.EyeRot.Forward;
+            forward += (Vector3.Random + Vector3.Random + Vector3.Random + Vector3.Random) * spread * 0.25f;
+            forward = forward.Normal;
+
+            foreach (TraceResult tr in TraceBullet(Owner.EyePos, Owner.EyePos + forward * 5000, bulletSize))
+            {
+                tr.Surface.DoBulletImpact(tr);
+
+                if (!tr.Entity.IsValid() || tr.Entity.IsWorld) return;
+                if (!IsServer)
+                {
+                    continue;
+                }
+
+                using (Prediction.Off())
+                {
+                    DamageInfo damageInfo = DamageInfo.FromBullet(tr.EndPos, forward * 100 * force, damage)
+                        .UsingTraceResult(tr)
+                        .WithAttacker(Owner)
+                        .WithWeapon(this);
+
+                    tr.Entity.TakeDamage(damageInfo);
+
+                    Vector3 pushDirection = new Vector3(-tr.Normal.x, -tr.Normal.y, tr.Normal.z < 0.5F ? 0.5F : -0.5F);
+                    float chargePercentage = MathX.Clamp((Time.Now - ChargingStartTime) / NEWTON_CHARGE_TIME, 0, 1);
+                    float chargeForce = ((NEWTON_FORCE_MAX - NEWTON_FORCE_MIN) * chargePercentage) + NEWTON_FORCE_MIN;
+
+                    tr.Entity.GroundEntity = null;
+                    tr.Entity.LocalVelocity = pushDirection * chargeForce;
```

Comment:
Do you know whether that's the way to apply velocity to an object? :o

Reply:
It is a way to apply velocity in gmod. I think in sbox it does the same.

Reply:
Yea just asking if there is another function that maybe adds some further calculations etc as setting the local velocity gonna kill the already applied velocity

Reply:
I don't think so but there is simple way to fix it.
`tr.Entity.LocalVelocity = tr.Entity.Velocity + pushDirection * chargeForce;`